### DISCUSSION TITLE
Investigation Segments

### DISCUSSION THREAD

**Discussion by ViMaSter** [Studio-Lovelies/GG-JointJustice-Unity#441] [Ideas]
More public notes than call-to-action, but wrote down some thoughts about a potential implementation of investigation sections.  
Written down just to hear other input though, nothing's set in stone!

---

Reference UI
![image](https://github.com/user-attachments/assets/c7637d1c-12c5-474c-815b-2bf922edcaa9)

---

Ink script could look something like…
```ink
&SCENE:Kitchen
&ACTOR:Dan
&MODE:Investigation
&ADD_FAILURE_SCRIPT:IRRELEVANT_EVIDENCE
+ [Arin being in the Kitchen]
    …dialogue…
    &UNLOCK_CHOICE:Waitrose LA branch
    …dialogue…
+ [Who bought the groceries?]
    …dialogue…
    &UNLOCK_SCENE:Waitrose
    …dialogue…
+ [LOCKED:Waitrose LA branch]
    …dialogue…
+ [Plumber_Invoice]
    …dialogue related to presented evidence…
```

Which would…
- load the `Kitchen` to the BGSceneList
- show Dan in the center (optional; if not added, only the room is shown)
- open the investigation menu (Examine, Move, Talk, Present) 

### Examine
- UI and actor are hidden and the player can click through the scene; the scene is also populated with `Detail`s
- like our `Actor` components, we also have GameObjects that have a new Component in scene prefabs: `Detail`
  - `Detail` takes a `DetailData` object containing properties:
    - 2D sprite
    - `IsPickup` (bool):
      - true: renderer and collider are disabled after first examination
      - false: remaining in the scene
    - `.ink` script that plays when examining it (which can either just play a dialogue or add evidence as before)
  - `Detail` has a SpriteRenderer and pixel-perfect collider so the player can "examine" it
  - after examining a detail, the `NarrativeScriptPlayer` stores `$SCENENAME_$DETAILNAME` (like `Kitchen_Post`) in `List<string> examinedDetails` and changes the cursor to indicate that this `Detail` was already examined (to indicate progression to the player)
    - `NarrativeScriptPlayer` instead of the `Detail`, as that state needs to be retained when `Move`ing between scenes

### Move
- allows moving to locations unlocked by `&UNLOCK_SCENE`

### Talk
- presents all `+ […]` options that **don't** match evidence names; similarly to [existing Choice sections](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/71eed67b992c56e7072344f6869e95a359e64dfe/unity-ggjj/Assets/Resources/InkDialogueScripts/Case1/1-13-TrialEnd.ink#L93-L100)
- `NarrativeScriptPlayer` tracks `List<string> examinedChoices` and adds `$SCENENAME_$CHOICETITLE` (like `Kitchen_Arin being in the Kitchen`) after selecting, to add a ✅ next to already read choices (to indicate progression to the player)
- Hides options prefixed with `LOCKED:` at first and only shows them after being unlocked by `&UNLOCK_CHOICE`

### Present
- identical to the existing ability to present evidence:
  - if a branch for that piece of evidence exists: play it
  - if not: fall back to `ADD_FAILURE_SCRIPT` entries

---

Considerations:
<details><summary>Why split out <code>Detail</code> from <code>Evidence</code>?</summary>
I think our current Evidence/Actor system is really robust in it's current form. This way, rather than changing it, we build on top: If a <code>Detail</code> adds <code>Evidence</code> to the court record, we can just do so in the <code>.ink</code> script associated with the <code>DetailData</code>.</br>  
Additionally not all details relate to evidence; writers want the ability to add details to a scene a player can interact with for optional dialogue that doesn't relate to game progress.<detail>
</details>
  
<details><summary>Should we add a prefix to clearly separate <code>Talk</code> and <code>Present</code> options?</summary>
No real reasoning here, just decided not to. I don't expect us to run into conflicts where we want to show a <code>Talk</code> option that shares the name with existing Evidence data.<br />
Happy to change it, if something like this feels more natural though:

```ink
&SCENE:Kitchen
&MODE:Investigation
&ADD_FAILURE_SCRIPT:IRRELEVANT_EVIDENCE
+ [TALK:Arin being in the Kitchen]
    …dialogue…
    &UNLOCK_CHOICE:Waitrose LA branch
    …dialogue…
+ [TALK:Who bought the groceries?]
    …dialogue…
    &UNLOCK_SCENE:Waitrose
    …dialogue…
+ [LOCKED:Waitrose LA branch]
    …dialogue…
+ [PRESENT:Plumber_Invoice]
    …dialogue related to presented evidence…
```

<detail>
</details>

**Comment by PhoebeMitchell** [comment]
My thoughts:

- Feels like the extra functionality proposed for NarrativeScriptPlayer here would be better suited for its own class specifically for managing that.
- There is functionality in ink for labeling choices which could maybe be used. e.g. `+ (locked) [Arin being in the Kitchen]` but not sure how easy it is to access that data programmatically.

Other than that this all makes sense to me :+1:

**Reply by ViMaSter** [reply]
> Feels like the extra functionality proposed for NarrativeScriptPlayer here would be better suited for its own class specifically for managing that.

_May_ have forgotten about the fact that we have the `NarrativeGameState` with references to almost everything. Decoupling it sounds way better! 👀  

Would add a `IInvestigationState`/`InvestigationState` instance which contains private `examinedChoices`/`examinedDetails` fields and public interfaces to manage them, unless you've got an even better place for it.

> There is functionality in ink for labeling choices which could maybe be used. e.g. + (locked) [Arin being in the Kitchen] but not sure how easy it is to access that data programmatically.

TIL. Sounds like the best way to go about it, if there's a way to interface with it, yeah. I'll check out the Ink docs and Unity integration again some time this week, see what comes up and post another update. (Unless someone else decides to check them out first! : D)
